libs/wingsdk/src/target-sim/function.inflight.ts

```diff
@@ -9,63 +13,176 @@ import {
     this.env = props.environmentVariables ?? {};
     this.context = context;
     this.timeout = props.timeout;
-    this.sandbox = new Sandbox(this.filename, {
-      env: {
-        ...this.env,
-        WING_SIMULATOR_URL: this.context.serverUrl,
-      },
-      timeout: this.timeout,
-      log: (internal, _level, message) => {
-        this.context.addTrace({
-          data: { message },
-          type: internal ? TraceType.RESOURCE : TraceType.LOG,
-          sourcePath: this.context.resourcePath,
-          sourceType: FUNCTION_FQN,
-          timestamp: new Date().toISOString(),
-        });
-      },
-    });
+    this.maxWorkers = props.concurrency;
+
+    this.createBundlePromise = this.createBundle();
+  }
+
+  // Create a bundle with the code that will be run by all workers.
+  private async createBundle() {
+    const workdir = await mkdtemp(path.join(tmpdir(), "wing-bundles-"));
+
+    let contents = (await readFile(this.originalFile)).toString();
+
+    // log a warning if contents includes __dirname or __filename
+    if (contents.includes("__dirname") || contents.includes("__filename")) {
+      this.addTrace(
+        `Warning: __dirname and __filename cannot be used within bundled cloud functions. There may be unexpected behavior.`,
+        false
+      );
+    }
+
+    // wrap contents with a shim that handles the communication with the parent process
+    // we insert this shim before bundling to ensure source maps are generated correctly
+    contents = `
+"use strict";
+${contents}
+process.on("message", async (message) => {
+  const { fn, args } = message;
+  try {
+    const value = await exports[fn](...args);
+    process.send({ type: "resolve", value });
+  } catch (err) {
+    process.send({ type: "reject", reason: err });
+  }
+});
+`;
+    const wrappedPath = this.originalFile.replace(/\.js$/, ".sandbox.js");
+    writeFileSync(wrappedPath, contents); // async fsPromises.writeFile "flush" option is not available in Node 20
+    const bundle = createBundle(wrappedPath, [], workdir);
+    this.bundledFile = bundle.entrypointPath;
+
+    if (process.env.DEBUG) {
+      const fileStats = await stat(this.originalFile);
+      this.addTrace(`Bundled code (${fileStats.size} bytes).`);
+    }
   }
 
   public async init(): Promise<FunctionAttributes> {
     return {};
   }
 
   public async cleanup(): Promise<void> {
-    await this.sandbox.cleanup();
+    await this.createBundlePromise;
+    for (const worker of this.workers) {
+      await worker.cleanup();
+    }
   }
 
   public async save(): Promise<void> {}
 
   public async invoke(payload: string): Promise<string> {
     return this.context.withTrace({
       message: `Invoke (payload=${JSON.stringify(payload)}).`,
-      activity: () => this.sandbox.call("handler", payload),
+      activity: async () => {
+        await this.createBundlePromise; // ensure inflight code is bundled before doing any work
```

Comment:
feels like this should be handled within the sandbox somehow... Maybe `SandboxBundle` holds this promise and the sandbox awaits it automatically?

Reply:
I think you're right... I moved over this responsibility of waiting for the bundle to be available from `invoke` to `initWorker`, it should be a little bit better